src/SymWoodburyMatrices.jl

```diff
@@ -0,0 +1,160 @@
+import Base:+,*,-,\,^,sparse,full,copy
+
+VectorTypes = Union{AbstractMatrix, Vector, SubArray, Diagonal}
+
+using Base.LinAlg.BLAS:gemm!,gemm
+
+"""
+Represents a matrix of the form A + BDBᵀ.
+"""
+type SymWoodbury{T,AType, BType, DType} <: AbstractMatrix{T}
+
+  A::AType; 
+  B::BType; 
+  D::DType;
+
+end
+
+"""
+    SymWoodbury(A, B, D)
+
+Represents a matrix of the form A + BDBᵀ.
+"""
+function SymWoodbury{T}(A, B::AbstractMatrix{T}, D)
+    n = size(A, 1)
+    k = size(B, 2)
+    if size(A, 2) != n || size(B, 1) != n || size(D,1) != k || size(D,2) != k
+        throw(DimensionMismatch("Sizes of B ($(size(B))) and/or D ($(size(D))) are inconsistent with A ($(size(A)))"))
+    end
+    SymWoodbury{T, typeof(A),typeof(B),typeof(D)}(A,B,D)
+end
+
+function calc_inv(A, B, D)
+
+  W = inv(A);
+  X = W*B;
+  invD = -inv(D);
+  Z = inv(invD - B'*X);
+  SymWoodbury(W,X,Z);
+
+end
+
+Base.inv{T<:Any, AType<:Any, BType<:Any, DType<:Matrix}(O::SymWoodbury{T,AType,BType,DType}) = 
+  calc_inv(O.A, O.B, O.D)
+
+# D is typically small, so this is acceptable. 
+Base.inv{T<:Any, AType<:Any, BType<:Any, DType<:SparseMatrixCSC}(O::SymWoodbury{T,AType,BType,DType}) = 
+  calc_inv(O.A, O.B, full(O.D));
+
+\(W::SymWoodbury, R::StridedVecOrMat) = inv(W)*R
+
+"""
+    partialInv(O)
+
+Get the factors (X,Z) in W + XZXᵀ where W + XZXᵀ = inv( A + BDBᵀ )
+"""
+function partialInv(O::SymWoodbury)
+
+  X = (O.A)\O.B;
+  invD = -1*inv(O.D);
+  Z = inv(invD - O.B'*X);
+  return (X,Z);
+
+end
+
+function liftFactorVars(A,B,Di)
+  n  = size(A,1)
+  k  = size(B,2)
+  M = [A    B   ;
+       B'  -Di ];
+  M = lufact(M) # ldltfact once it's available.
+  return x -> (M\[x; zeros(k,1)])[1:n,:];
+end
+
+"""
+    liftFactor(A)
+
+More stable version of inv(A).  Returns a function which computs the inverse
+on evaluation, i.e. `liftFactor(A)(x)` is the same as `inv(A)*x`.
+"""
+liftFactor{T<:Any,
+           AType<:Any, 
+           BType<:Any, 
+           DType<:Matrix}(O::SymWoodbury{T,AType,BType,DType}) = 
+           liftFactorVars(sparse(O.A), sparse(O.B), sparse(inv(O.D))); 
+
+liftFactor{T<:Any,
+           AType<:Any, 
+           BType<:Any, 
+           DType<:SparseMatrixCSC}(O::SymWoodbury{T,AType,BType,DType}) = 
+           liftFactorVars(sparse(O.A), sparse(O.B), sparse(inv(full(O.D))));
+
+# Optimization - use specialized BLAS package 
+function *{T<:Any, AType<:Any, BType<:Matrix, DType<:Any}(O::SymWoodbury{T,AType,BType,DType}, 
+  x::Array{Float64,2})
```

Comment:
Are you especially interested in the two-dimensionality or the `Float64`? Likewise, aren't there constraints on the eltypes of `O.B` and `o`? Your best bet might be to define `*` as something that does all necessary pre-computations (e.g., `o = O.A*x` etc) and then calls some other function; you can use dispatch on that function to control whether you use the BLAS algorithm or not.

Also, handy tip:
```jl
julia> Base.LinAlg.BlasFloat
Union{Complex{Float32},Complex{Float64},Float32,Float64}
```